src/core/mesh/qgsmeshdataprovider.h

```diff
@@ -227,6 +227,93 @@ class CORE_EXPORT QgsMeshDataBlock
     DataType mType;
 };
 
+/**
+ * \ingroup core
+ *
+ * QgsMesh3dDataBlock is a block of 3d stacked mesh data related N
+ * faces defined on base mesh frame.
+ *
+ * data are implicitly shared, so the class can be quickly copied
+ * std::numeric_limits<double>::quiet_NaN() represents NODATA value
+ *
+ * \since QGIS 3.12
+ */
+class CORE_EXPORT QgsMesh3dDataBlock
+{
+  public:
+    //! Type of data stored in the block
+    enum DataType
+    {
+      ActiveFlagInteger, //!< Integer boolean flag whether volume is active
+      ScalarDouble, //!< Scalar double values
+      VectorDouble, //!< Vector double pairs (x1, y1, x2, y2, ... )
+      VerticalLevelsCount, //! Integer number of vertical levels above face
+      VerticalLevels, //! Double - size + 1
+      FaceToVolumeIndex //! Integer
+    };
+
+    //! Constructs an invalid block
+    QgsMesh3dDataBlock();
+
+    //! Dtor
+    ~QgsMesh3dDataBlock();
+
+    //! Constructs a new block for count faces
+    QgsMesh3dDataBlock( int count, int maximumVerticalLevels, bool isVector );
+
+    //! Sets block as valid
+    void setIsValid();
```

Comment:
can we rename to setValid( bool valid ) to make it consistent with other Qt APIs?